supervisor/homeassistant/websocket.py

```diff
@@ -22,41 +26,95 @@ class WSClient:
+        """Close down the client."""
+        if not self._client.closed:
+            await self._client.close()
+
+    async def async_send_command(self, message: dict[str, Any]) -> dict | None:
+        """Send a websocket message, and return the response."""
+        self._message_id += 1
+        message["id"] = self._message_id
+        self._futures[message["id"]] = self._loop.create_future()
+        _LOGGER.debug("Sending: %s", message)
+        try:
+            await self._client.send_json(message)
+        except ConnectionError as err:
+            raise HomeAssistantWSConnectionError(err) from err
 
-    async def async_send_command(self, message: dict[str, Any]):
-        """Send a websocket command."""
-        async with self._lock:
-            self.message_id += 1
-            message["id"] = self.message_id
+        try:
+            return await self._futures[message["id"]]
+        finally:
+            self._futures.pop(message["id"])
 
-            _LOGGER.debug("Sending: %s", message)
-            try:
-                await self.client.send_json(message)
-            except ConnectionError:
-                return
+    async def start_listener(self) -> None:
+        """Start listening to the websocket."""
+        if not self.connected:
+            raise HomeAssistantWSConnectionError("Not connected when start listening")
 
-            try:
-                response = await self.client.receive_json()
-            except ConnectionError:
-                return
+        try:
+            while self.connected:
+                await self._receive_json()
+        except HomeAssistantWSConnectionError:
+            pass
+
+        except HomeAssistantWSError as err:
+            _LOGGER.warning(err)
+
+        finally:
+            await self.close()
+
+    async def _receive_json(self) -> None:
+        """Receive json."""
+        msg = await self._client.receive()
+        _LOGGER.debug("Received: %s", msg)
 
-            _LOGGER.debug("Received: %s", response)
+        if msg.type in (
+            WSMsgType.CLOSE,
+            WSMsgType.CLOSED,
+            WSMsgType.CLOSING,
+            WSMsgType.ERROR,
+        ):
+            raise HomeAssistantWSConnectionError()
+
+        if msg.type != WSMsgType.TEXT:
+            raise HomeAssistantWSError(f"Received non-Text message: {msg.type}")
+
+        try:
+            data = msg.json()
+        except ValueError as err:
+            raise HomeAssistantWSError(f"Received invalid JSON - {msg}") from err
 
-            if response["success"]:
-                return response["result"]
+        if data["type"] == "result":
+            if (future := self._futures.get(data["id"])) is None:
+                return
+
+            if data["success"]:
+                future.set_result(data["result"])
+                return
 
-            raise HomeAssistantWSError(response)
+            _LOGGER.error("Unsuccessful websocket message - %s", data)
```

Comment:
But I think in this case we could inform the future about the unsuccessful message. I guess setting an exception would be most fitting here.

Reply:
Yeah, we could do that :+1: